lib/goal/container/container.ts

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright © 2019 Atomist, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+    GitProject,
+    HandlerContext,
+} from "@atomist/automation-client";
+import {
+    DefaultGoalNameGenerator,
+    FulfillableGoalDetails,
+    FulfillableGoalWithRegistrations,
+    Fulfillment,
+    getGoalDefinitionFrom,
+    Goal,
+    GoalFulfillmentCallback,
+    ImplementationRegistration,
+    SdmGoalEvent,
+    SoftwareDeliveryMachine,
+} from "@atomist/sdm";
+import {
+    isConfiguredInEnv,
+    KubernetesGoalScheduler,
+} from "../../pack/k8s/KubernetesGoalScheduler";
+import { KubernetesJobDeletingGoalCompletionListenerFactory } from "../../pack/k8s/KubernetesJobDeletingGoalCompletionListener";
+import { toArray } from "../../util/misc/array";
+import {
+    CacheEntry,
+    cachePut,
+    cacheRestore,
+} from "../cache/goalCaching";
+import { dockerContainerScheduler } from "./docker";
+import { k8sContainerScheduler } from "./k8s";
+import { runningInK8s } from "./util";
+
+/**
+ * Create and return a container goal with the provided container
+ * specification.
+ *
+ * @param displayName Goal display name
+ * @param registration Goal containers, volumes, cache details, etc.
+ * @return SDM container goal
+ */
+export function containerGoal(displayName: string, registration: ContainerRegistration): Container {
```

Comment:
`containerGoal` as the name of the function feels inconsistent from the rest of SDM. We normally don't use the `goal` suffix for goal classes or creation functions. The jenkins goal is called `Jenkins` and the creation function `jenkinsRun`.

Was there any particular reason to use `containerGoal` instead of `container` or `containerRun`?

Reply:
I'm happy to shorten it to `container`.

Reply:
I called the function `containerGoal` because it returns a Container goal.

Reply:
Cool, let's go with `container` then. And to be consistent, I'll rename the `jenkinsRun` to `jenkins`.